shadow-utils-4.1.4.2/man/expiry.1
'\" t
.\"     Title: expiry
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.1 <http://docbook.sf.net/>
.\"      Date: 07/24/2009
.\"    Manual: User Commands
.\"    Source: User Commands
.\"  Language: English
.\"
.TH "EXPIRY" "1" "07/24/2009" "User Commands" "User Commands"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
expiry \- check and enforce password expiration policy
.SH "SYNOPSIS"
.HP \w'\fBexpiry\fR\ 'u
\fBexpiry\fR [\-c] [\-f]
.SH "DESCRIPTION"
.PP
The
\fBexpiry\fR
command checks (\fB\-c\fR) the current password expiration and forces (\fB\-f\fR) changes when required\&. It is callable as a normal user command\&.
.SH "FILES"
.PP
/etc/passwd
.RS 4
User account information\&.
.RE
.PP
/etc/shadow
.RS 4
Secure user account information\&.
.RE
.SH "SEE ALSO"
.PP

\fBpasswd\fR(5),
\fBshadow\fR(5)\&.


gmime-38b35d8/ChangeLog
2015-01-23  Sunghyun Kwon     <sh0701.kwon@samsung.com>

    * gmime/gmime-gpg-context.c (gpg_ctx_set_mode): Added the condition
    (GPG_CTX_MODE_SIGN_ENCRYPT) 

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.15

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-html.c (html_convert): Fixed to convert the
	last line of input even if it doesn't end with a \n.

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.14

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* util/url-scanner.c (url_addrspec_start): Fixed the logic to
	actually scan for the beginning of the mailbox.

2012-12-09  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.13

2012-11-26  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_iter_first): Make sure the
	list isn't empty.
	(g_mime_header_iter_last): Same.

2012-11-23  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_list_prepend): Prepend the
	header to the list, don't *append* it. Oops.
	(g_mime_header_list_clear): Don't forget to reset the linked list
	so that it doesn't contained free'd nodes.

2012-11-02  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.12

2012-10-27  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-stream-filter.c (stream_read): Initialize presize to
	READ_PAD instead of READ_SIZE to prevent potential buffer
	underruns.

2012-10-20  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-gzip.c (gunzip_filter): Apparently
	Z_BUF_ERROR isn't actually an error unless there is input
	available. Thanks to Pavlo <devgs@ukr.net> for this patch.

2012-10-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-parser.c (parser_push_boundary): Don't use strcmp to
	check if the boundary is an mbox boundary, use a direct pointer
	comparison so that we can handle Content-Type boundary parameters
	with a value of "From "

2012-10-19  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-parser.c (parser_step_headers): Don't assume that
	the first non-header line we find indicates the start of the
	content unless we've already gotten some basic headers.

	Fix for bug #686223.

2012-10-18  Jeffrey Stedfast  <jeff@xamarin.com>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.11

2012-09-12  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Unref the
	certificate after adding it to the list (adding it to the list
	refs it).

	* gmime/gmime-multipart-encrypted.c
	(g_mime_multipart_encrypted_decrypt): If the 'result' out param is
	null, unref the GMimeDecryptResult to avoid leaking.

2012-08-13  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (tokenize_rfc2047_phrase): Don't forget to
	initialize 'ascii' to TRUE.

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-stream-pipe.c (stream_flush): Pipes and sockets
	cannot be fsync()'d, so don't even try!

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (rfc2047_token_new_encoded_word): Make sure
	to properly handle broken encoded-word tokens in the form:
	=?charset?q?=.

	Fixes the g_malloc abort() in bug #679388

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-multipart-signed.c (check_protocol_supported): Oops,
	need to compare against xsupported in the second string compare.

	Fixes bug #674032

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.10

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/Makefile.am: Always build gmime-gpg-context.[c,h] and
	gmime-pkcs7-context.[c,h]

	* gmime/gmime-gpg-context.c: Armor internals with #ifdef
	ENABLE_CRYPTOGRAPHY so that if crypto is disabled, only the
	skeletal API is built.

	* gmime/gmime.h: Include gmime-pkcs7-context.h

	Fixes bug #676692.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): Use
	is_blank() instead of is_lwsp() so that we only treat space and
	tab as word delimiters (we want to rfc2047 encode \r's and
	\n's). Check for ctrl characters when deciding what the type and
	encoding of the word is.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): When
	breaking up really long atom tokens, change the word type to
	RFC2047 so that we encode them, thus preventing forced header
	folding which cannot be unmunged.
	(header_fold): Fixed for cases where the first word in a header is
	68 chars. (Note: this fix is largely unnecessary due to the above fix)

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-message.c: When setting a header, make sure to
	always clear the header list cache of the toplevel mime_part as
	well.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-param.c (rfc2184_param_new): Only decode the charset
	and lang if the parameter is actually encoded. Thanks to Peter
	Bloomfield for this patch.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Keep track of
	DECRYPTION_OKAY status.
	(gpg_decrypt): Only return NULL if decryption failed (e.g. we
	didn't get a DECRYPTION_OKAY status).

	Fixes bug #677088.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* docs/reference/building.sgml: Updated docs.

	Fixes bug #676341

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.9

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-multipart-signed.c (g_mime_multipart_signed_verify):
	Match protocols like application/x-pkcs7-signature, too.

	Fixes bug #674032.

2012-04-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.8


shadow-utils-4.1.4.2/src/expiry.c
/*
 * Copyright (c) 1994       , Julianne Frances Haugh
 * Copyright (c) 1996 - 2000, Marek Michałkiewicz
 * Copyright (c) 2001 - 2006, Tomasz Kłoczko
 * Copyright (c) 2007 - 2008, Nicolas François
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the copyright holders or contributors may not be used to
 *    endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: expiry.c 2650 2009-04-11 18:37:08Z nekral-guest $"

#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <sys/types.h>
#include "defines.h"
#include "prototypes.h"

/* Global variables */
char *Prog;

/* local function prototypes */
static RETSIGTYPE catch_signals (int);
static void usage (void);

/*
 * catch_signals - signal catcher
 */
static RETSIGTYPE catch_signals (unused int sig)
{
	exit (10);
}

/*
 * usage - print syntax message and exit
 */
static void usage (void)
{
	fputs (_("Usage: expiry {-f|-c}\n"), stderr);
	exit (10);
}

/* 
 * expiry - check and enforce password expiration policy
 *
 *	expiry checks (-c) the current password expiration and forces (-f)
 *	changes when required. It is callable as a normal user command.
 */
int main (int argc, char **argv)
{
	struct passwd *pwd;
	struct spwd *spwd;

	Prog = Basename (argv[0]);

	sanitize_env ();

	/* 
	 * Start by disabling all of the keyboard signals.
	 */
	(void) signal (SIGHUP, catch_signals);
	(void) signal (SIGINT, catch_signals);
	(void) signal (SIGQUIT, catch_signals);
#ifdef	SIGTSTP
	(void) signal (SIGTSTP, catch_signals);
#endif

	/*
	 * expiry takes one of two arguments. The default action is to give
	 * the usage message.
	 */
	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

	OPENLOG ("expiry");

	if (   (argc != 2)
	    || (   (strcmp (argv[1], "-f") != 0)
	        && (strcmp (argv[1], "-c") != 0))) {
		usage ();
	}

	/*
	 * Get user entries for /etc/passwd and /etc/shadow
	 */
	pwd = get_my_pwent ();
	if (NULL == pwd) {
		fprintf (stderr, _("%s: Cannot determine your user name.\n"),
		         Prog);
		SYSLOG ((LOG_WARN, "Cannot determine the user name of the caller (UID %lu)",
		         (unsigned long) getuid ()));
		exit (10);
	}
	spwd = getspnam (pwd->pw_name); /* !USE_PAM, No need for xgetspnam */

	/*
	 * If checking accounts, use agecheck() function.
	 */
	if (strcmp (argv[1], "-c") == 0) {

		/*
		 * Print out number of days until expiration.
		 */
		agecheck (spwd);

		/*
		 * Exit with status indicating state of account.
		 */
		exit (isexpired (pwd, spwd));
	}

	/*
	 * If forcing password change, use expire() function.
	 */
	if (strcmp (argv[1], "-f") == 0) {

		/*
		 * Just call expire(). It will force the change or give a
		 * message indicating what to do. And it doesn't return at
		 * all unless the account is unexpired.
		 */
		expire (pwd, spwd);
		exit (0);
	}

	/*
	 * Can't get here ...
	 */
	usage ();
	exit (1);
}

